File: internship_backend_apr25/jury_solutions/java/A.java
import java.util.HashSet;
import java.util.Scanner;

public class A {
	public static void main(String[] args) throws Exception {
		new A().run();
	}

	private final int AllCardsCount = 52;

	private final HashSet<Character> AllSuits = new HashSet<>() {{
		add('C');
		add('D');
		add('H');
		add('S');
	}};

	private final HashSet<Character> AllRanks = new HashSet<>() {{
		add('2');
		add('3');
		add('4');
		add('5');
		add('6');
		add('7');
		add('8');
		add('9');
		add('T');
		add('J');
		add('Q');
		add('K');
		add('A');
	}};

	private Scanner scanner;

	public void run() {
		scanner = new Scanner(System.in);
		String[] groupsCount = scanner.nextLine().split(" ");
		double probability1 = calculateProbability(Integer.parseInt(groupsCount[0]), Integer.parseInt(groupsCount[1]));
		double probability2 = calculateProbability(Integer.parseInt(groupsCount[2]), Integer.parseInt(groupsCount[3]));

		System.out.println(Math.max(probability1, probability2));
	}

	private double calculateProbability(int removedGroupsCount, int soughtGroupsCount) {
		HashSet<String> removedCards = new HashSet<>();
		for (int i = 0; i < removedGroupsCount; i++) {
			HashSet<Character> group = new HashSet<>();
			for (char c : scanner.nextLine().toCharArray())
				group.add(c);

			removedCards.addAll(getCards(group));
		}

		HashSet<String> soughtCards = new HashSet<>();
		for (int i = 0; i < soughtGroupsCount; i++) {
			HashSet<Character> group = new HashSet<>();
			for (char c : scanner.nextLine().toCharArray())
				group.add(c);
			soughtCards.addAll(getCards(group));
		}

		if (removedCards.size() == AllCardsCount)
			return 0.0;

		int removedCardsCount = removedCards.size();

		removedCards.retainAll(soughtCards);
		int cardsIntersection = removedCards.size();

		return (double) (soughtCards.size() - cardsIntersection) / (AllCardsCount - removedCardsCount);
	}

	private HashSet<String> getCards(HashSet<Character> group) {
		HashSet<Character> currentSuits = new HashSet<>(AllSuits);
		currentSuits.retainAll(group);
		if (currentSuits.isEmpty())
			currentSuits = new HashSet<>(AllSuits);

		group.removeAll(currentSuits);
		HashSet<Character> currentRanks = group;
		if (currentRanks.isEmpty())
			currentRanks = new HashSet<>(AllRanks);

		HashSet<String> cards = new HashSet<>();
		for (Character rank : currentRanks) {
			for (Character suit : currentSuits)
				cards.add(rank.toString() + suit);
		}

		return cards;
	}
}

File: internship_backend_apr25/jury_solutions/go/a.go
package main

import (
	"fmt"
	"maps"
	"strconv"
)
import "os"
import "bufio"

var AllSuits = map[rune]struct{}{
	'C': {},
	'D': {},
	'H': {},
	'S': {},
}

var AllRanks = map[rune]struct{}{
	'2': {},
	'3': {},
	'4': {},
	'5': {},
	'6': {},
	'7': {},
	'8': {},
	'9': {},
	'T': {},
	'J': {},
	'Q': {},
	'K': {},
	'A': {},
}

var AllCardsCount = 52

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 1000000), 1000000)

	var r1, s1, r2, s2 int
	scanner.Scan()
	fmt.Sscan(scanner.Text(), &r1, &s1, &r2, &s2)
	probability1 := CalculateProbability(scanner, r1, s1)
	probability2 := CalculateProbability(scanner, r2, s2)

	if probability1 > probability2 {
		fmt.Println(probability1)
	} else {
		fmt.Println(probability2)
	}
}

func CalculateProbability(scanner *bufio.Scanner, r int, s int) float64 {
	removedCards := make(map[string]struct{})
	for i := 0; i < r; i++ {
		scanner.Scan()
		group := scanner.Text()
		_ = group
		maps.Copy(removedCards, GetCards(group))
	}

	soughtCards := make(map[string]struct{})
	for i := 0; i < s; i++ {
		scanner.Scan()
		group := scanner.Text()
		_ = group
		maps.Copy(soughtCards, GetCards(group))
	}
	
	if len(removedCards) == AllCardsCount {
		return 0
	}
	var cardsIntersection int
	for card, _ := range removedCards {
		_, ok := soughtCards[card]
		if ok {
			cardsIntersection++
		}
	}
	return float64(len(soughtCards)-cardsIntersection) / float64(AllCardsCount-len(removedCards))
}

func GetCards(group string) map[string]struct{} {
	currentSuits := make(map[rune]struct{})
	for _, suit := range group {
		_, ok := AllSuits[suit]
		if ok {
			currentSuits[suit] = struct{}{}
		}
	}
	if len(currentSuits) == 0 {
		currentSuits = AllSuits
	}

	currentRanks := make(map[rune]struct{})
	for _, rank := range group {
		_, ok := AllRanks[rank]
		if ok {
			currentRanks[rank] = struct{}{}
		}
	}
	if len(currentRanks) == 0 {
		currentRanks = AllRanks
	}

	cards := make(map[string]struct{})
	for suit, _ := range currentSuits {
		for rank, _ := range currentRanks {
			var card = strconv.QuoteRune(suit) + strconv.QuoteRune(rank)
			cards[card] = struct{}{}
		}
	}

	return cards
}
